### PR TITLE
Fixed up the version mask rules

### PR DESCRIPTION
**Pull request by mcasperson**
The masking rules incorrectly interpreted a version like `0.1.0-initial` as a mask for the revision.

This PR changes the masking rules to be more strict, with only dots allowed between version components and a dash before the prerelease. This change removed the ambiguity of what can be a masked component, while still producing the same masked version results.

### CODE REVIEW

**Comment by mcasperson** [comment]
[Releases version number created in Octopus are being modified and created incorrectly](https://trello.com/c/u06eZg7t/3920-releases-version-number-created-in-octopus-are-being-modified-and-created-incorrectly)

File: source/Octopus.Versioning/Octopus/OctopusVersionMask.cs [hunk 1]
```diff
@@ -188,7 +188,7 @@ public override string EvaluateFromMask(string separator = "")
                     if (!IsPresent || string.IsNullOrEmpty(Value))
                         return string.Empty;
 
-                    var identifiers = Value.Split('.', '-', '_');
+                    var identifiers = Value.Split('.');
```
Comment (mcasperson) [inline source/Octopus.Versioning/Octopus/OctopusVersionMask.cs:191]: This, along with the changes in OctopusVersionMaskParser.cs below, revert back to a more strict interpretation of masks.

File: source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs [hunk 2]
```diff
@@ -21,27 +21,27 @@ public class OctopusVersionMaskParser
             // Get the major version number
             @$"(?<{Major}>\d+|{PatternIncrement}|{PatternCurrent})" +
             // Get the minor version number, delimited by a period, comma, dash or underscore
-            @$"(?:[.\-_](?<{Minor}>\d+|{PatternIncrement}|{PatternCurrent}))?" +
+            @$"(?:\.(?<{Minor}>\d+|{PatternIncrement}|{PatternCurrent}))?" +
```
Comment (mcasperson) [inline source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs:24]: These changes make the masks more strict, and align them with the output version generated by a mask, which has dots between version components and dash before the prerelease.

File: source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs [hunk 3]
```diff
@@ -21,27 +21,27 @@ public class OctopusVersionMaskParser
...
             // Get the patch version number, delimited by a period, comma, dash or underscore
-            @$"(?:[.\-_](?<{Patch}>\d+|{PatternIncrement}|{PatternCurrent}))?" +
+            @$"(?:\.(?<{Patch}>\d+|{PatternIncrement}|{PatternCurrent}))?" +
             // Get the revision version number, delimited by a period, comma, dash or underscore
-            @$"(?:[.\-_](?<{Revision}>\d+|{PatternIncrement}|{PatternCurrent}))?)?" +
+            @$"(?:\.(?<{Revision}>\d+|{PatternIncrement}|{PatternCurrent}))?)?" +
             // Everything after the last digit and before the plus is the prerelease
-            @$"(?:[.\-_])?(?<{Prerelease}>(?<{PrereleasePrefix}>[^+.\-_\s]*?)([.\-_](?<{PrereleaseCounter}>[^+\s]*?)?)?)?" +
+            @$"(?:-(?<{Prerelease}>(?<{PrereleasePrefix}>[^+.\-_\s]*?)([.\-_](?<{PrereleaseCounter}>[^+\s]*?)?)?)?)?" +
```
Comment (mcasperson) [inline source/Octopus.Versioning/Octopus/OctopusVersionMaskParser.cs:30]: This is the change that fixes the reported issue. It was here that a prerelease that started with an 'i' or 'c' was counted as a mask, where it wasn't previously.

File: source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs [hunk 4]
```diff
@@ -11,40 +11,89 @@ public class OctopusVersionMaskParserTests
         static readonly OctopusVersionMaskParser OctopusVersionMaskParser = new OctopusVersionMaskParser();
         static readonly OctopusVersionParser OctopusVersionParser = new OctopusVersionParser();
 
+        /// <summary>
+        /// This test used to live in Octopus server. You can find it at commit 43e42a5769083dffa65e2c9ab4d513641c37248a
+        /// in the SemanticVersionMaskFixture.cs file.
+        ///
+        /// Tests here, with the exception of versions that start with a 'v' or 'V', should also pass in
+        /// that old commit.
+        /// </summary>
```
Comment (mcasperson) [inline source/Octopus.Versioning.Tests/Octopus/OctopusVersionMaskParserTests.cs:20]: Added a bunch more tests to capture the edge case that resulted in this PR.
